File: data/src/main/java/com/everis/data/network/firebase/FireBaseAuthImpl.java
package com.everis.data.network.firebase;

import android.support.annotation.NonNull;

import com.everis.data.model.FirebaseUserDataMapper;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import rx.Observable;
import rx.Subscriber;

/**
 * Created by everis on 25/04/18.
 */

public class FireBaseAuthImpl implements FireBaseAuth {

    @Override
    public Observable<FirebaseUser> createUser(final String email, final String password) {
        return Observable.create(new Observable.OnSubscribe<FirebaseUser>() {
            @Override
            public void call(final Subscriber<? super FirebaseUser> subscriber) {
                FirebaseAuth.getInstance().createUserWithEmailAndPassword(email, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if(task.isSuccessful()){
                            subscriber.onNext(task.getResult().getUser());
                        }
                        else{
                            subscriber.onError(new Exception());
                        }
                        subscriber.onCompleted();
                    }
                });
            }
        });
    }

    @Override
    public Observable<FirebaseUser> signInUser(final String emaail, final String password) {
        return Observable.create(new Observable.OnSubscribe<FirebaseUser>() {
            @Override
            public void call(final Subscriber<? super FirebaseUser> subscriber) {
                FirebaseAuth.getInstance().signInWithEmailAndPassword(emaail,password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if(task.isSuccessful()){
                            subscriber.onNext(task.getResult().getUser());
                            subscriber.onCompleted();
                        }
                        else{
                             subscriber.onError(task.getException());
                        }
                    }
                });
            }
        });
    }

    @Override
    public Observable<FirebaseUser> observeAuthStateChange() {
        return Observable.create(new Observable.OnSubscribe<FirebaseUser>() {
            @Override
            public void call(final Subscriber<? super FirebaseUser> subscriber) {
                FirebaseAuth.getInstance().addIdTokenListener(new FirebaseAuth.IdTokenListener() {
                    @Override
                    public void onIdTokenChanged(@NonNull FirebaseAuth firebaseAuth) {
                        subscriber.onNext(firebaseAuth.getCurrentUser());

                    }
                });
            }
        });
    }

    @Override
    public Observable<Void> logOut() {
        return Observable.create(new Observable.OnSubscribe<Void>() {
            @Override
            public void call(Subscriber<? super Void> subscriber) {
                FirebaseAuth.getInstance().signOut();
                subscriber.onCompleted();
            }
        });
    }

    @Override
    public Observable<Void> deleteUser() {
        return Observable.create(new Observable.OnSubscribe<Void>() {
            @Override
            public void call(final Subscriber<? super Void> subscriber) {
                if(FirebaseAuth.getInstance().getCurrentUser()!=null){
                    FirebaseAuth.getInstance().getCurrentUser().delete().addOnCompleteListener(new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) {
                            if(task.isSuccessful()){
                                subscriber.onCompleted();
                            }
                            else{
                                subscriber.onError(task.getException());
                            }
                        }
                    });
                }
            }
        });
    }

    @Override
    public Observable<FirebaseUser> checkUserLoggedIn() {
        return Observable.create(new Observable.OnSubscribe<FirebaseUser>() {
            @Override
            public void call(Subscriber<? super FirebaseUser> subscriber) {
                subscriber.onNext(FirebaseAuth.getInstance().getCurrentUser());
                subscriber.onCompleted();
            }
        });
    }
}

File: data/src/main/java/com/everis/data/network/firebase/FireBaseAuthImpl.kt
package com.everis.data.network.firebase

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser

import rx.Observable

class FireBaseAuthImpl : FireBaseAuth {

    override fun createUser(email: String, password: String): Observable<FirebaseUser> {
        return Observable.create { subscriber ->
            FirebaseAuth.getInstance().createUserWithEmailAndPassword(email, password).addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    subscriber.onNext(task.result.user)
                } else {
                    subscriber.onError(Exception())
                }
                subscriber.onCompleted()
            }
        }
    }

    override fun signInUser(email: String, password: String): Observable<FirebaseUser> {
        return Observable.create { subscriber ->
            FirebaseAuth.getInstance().signInWithEmailAndPassword(email, password).addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    subscriber.onNext(task.result.user)
                    subscriber.onCompleted()
                } else {
                    subscriber.onError(task.exception)
                }
            }
        }
    }

    override fun observeAuthStateChange(): Observable<FirebaseUser> {
        return Observable.create { subscriber -> FirebaseAuth.getInstance().addIdTokenListener { firebaseAuth -> subscriber.onNext(firebaseAuth.currentUser) } }
    }

    override fun logOut(): Observable<Void> {
        return Observable.create { subscriber ->
            FirebaseAuth.getInstance().signOut()
            subscriber.onCompleted()
        }
    }

    override fun deleteUser(): Observable<Void> {
        return Observable.create { subscriber ->
            if (FirebaseAuth.getInstance().currentUser != null) {
                FirebaseAuth.getInstance().currentUser!!.delete().addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        subscriber.onCompleted()
                    } else {
                        subscriber.onError(task.exception)
                    }
                }
            }
        }
    }

    override fun checkUserLoggedIn(): Observable<FirebaseUser> {
        return Observable.create { subscriber ->
            subscriber.onNext(FirebaseAuth.getInstance().currentUser)
            subscriber.onCompleted()
        }
    }
}
